### DISCUSSION TITLE
quarkus-http connector

### DISCUSSION THREAD

**Discussion by luca-bassoricci** [quarkusio/quarkus#34221] [Q&A] [answered]
I'm using quarkus-http reactive messaging connector.
Message is emitted using `Emitter`.
No resource class (annotated with @Path) has been created because I have intended that "virtual" endpoints are automatically exposed.
Service1 receive messages from Service2 via messages:
```
## Service1
mp.messaging.incoming.incoming-mail.connector=quarkus-http
mp.messaging.incoming.incoming-mail.path=/mail-collector
mp.messaging.incoming.incoming-mail.method=POST
```
on the other side (Service2)
```
mp.messaging.outgoing.outgoing-result-mail.connector=quarkus-http
mp.messaging.outgoing.outgoing-result-mail.url=http://service1:8080/mail-collector
mp.messaging.outgoing.outgoing-result-mail.method=POST
```
all is working fine in local but when moving to openshift nothing happens.
Services are deployed on the same OS cluster as separate services; I'am sure they can communicate because we are doing some call using restclient reactive and everything is fine.
Using cURL Service1 host is pinged but endpoint exposed as channel (http://service1:8080/mail-collector) return 404.
Latest Quarkus 2.x version. 
Any advice?

(sorry for small info, but I'm in rush now).

[a comment by an automated account is not shown]

**Reply by geoand** [reply]
cc @cescoffier

**Reply by geoand** [reply]
Are there any errors or warnings in the logs that could point to a problem?

**Reply by luca-bassoricci** [reply]
I TRACEd and nothing special is logged, just
```
2023-06-22 08:47:29 DEBUG pod=cloud-7674c65d45-hcflx, traceId=, parentId=, spanId=, sampled= [io.sm.re.me.provider] (main) SRMSG00226: Found incoming connectors: [quarkus-websocket, quarkus-http]
2023-06-22 08:47:29 DEBUG pod=cloud-7674c65d45-hcflx, traceId=, parentId=, spanId=, sampled= [io.sm.re.me.provider] (main) SRMSG00227: Found outgoing connectors: [quarkus-websocket, quarkus-http]
2023-06-22 08:47:29 DEBUG pod=cloud-7674c65d45-hcflx, traceId=, parentId=, spanId=, sampled= [io.sm.re.me.provider] (main) SRMSG00229: Channel manager initializing...
2023-06-22 08:47:29 DEBUG pod=cloud-7674c65d45-hcflx, traceId=, parentId=, spanId=, sampled= [io.sm.re.me.provider] (main) SRMSG00209: Beginning graph resolution, number of components detected: 4
2023-06-22 08:47:29 DEBUG pod=cloud-7674c65d45-hcflx, traceId=, parentId=, spanId=, sampled= [io.sm.re.me.provider] (main) SRMSG00210: Graph resolution completed in 1972727 ns
2023-06-22 08:47:29 DEBUG pod=cloud-7674c65d45-hcflx, traceId=, parentId=, spanId=, sampled= [io.sm.re.me.provider] (main) SRMSG00235: Beginning materialization
2023-06-22 08:47:29 DEBUG pod=cloud-7674c65d45-hcflx, traceId=, parentId=, spanId=, sampled= [io.sm.re.me.provider] (main) SRMSG00236: Materialization completed in 14914322 ns
```

added to OP: 
Message is emitted using `Emitter`.

**Comment by luca-bassoricci** [comment]
I resolved removing `@Incoming` channel configuration and replace with "real" endpoint like:
```
@Path("")
class IncomingReplacementeResource {
  @Path("/mail-collector")
  @POST
  @Consumes("application/json")
  public void mailCollector(BodyDTO body, @Context UriInfo uriInfo) {
    // extract data from uriinfo, headers and so on as replacement of `IncomingHttpMetadata`
  }
}
```
On the "caller" side I'm still using an outgoing channel and everything works fine,but it's just a workaround.

**Comment by cescoffier** [comment]
I tried to reproduce it, but it seems to work OOTB.

The only difference is:
`mp.messaging.outgoing.outgoing-result-mail.url=http://service1:80/mail-collector` => the (generated) service is listening on the port 80 not 8080, 


I built the 2 apps and then used `quarkus deploy openshift --image-build`, and everything worked. 

Can you give me more details about the issue?

**Reply by luca-bassoricci** [reply]
@cescoffier thanks for you time.
I'm doing some tests using different ports than 8080: on local machine the 2 services are listening on different ports to avoid conflicts but on OS they are deployed on the same port (8080). I don't think this will be the problem, but I have time to give it a try.

**Comment by luca-bassoricci** [comment, marked as answer]
Solved!
I didn't configure any `mp.messaging.incoming.incoming-mail.*` in project 'application.properties' because I configure them directly in Openshift "service deployment" section and seems that `mp.messaging.incoming` properties setted from environment are not correctly managed.
I moved `mp.messaging.incoming.*` to master `application.properties` on my project and all is working fine.
I don't known if beans holding `@Incoming()` annotation are removed during build because no `mp.messaging.incoming` properties are found in configuration(s) or if `mp.messaging.incoming` properties are build-time only properties and I was misused them.

(https://github.com/quarkiverse/quarkus-reactive-messaging-http/blob/00a9f4bdd9a13f35244720b562dfab7ac544f9b4/deployment/src/main/java/io/quarkus/reactivemessaging/http/deployment/ReactiveHttpProcessor.java#L89): 
```
@BuildStep
@Record(ExecutionTime.RUNTIME_INIT)
void registerHttpConnector(...
```
I was checking extension code and I was convinced `mp.messaging.incoming` were runtime properties because `RUNTIME_INIT` executiontime.

Thanks to all you guys for your time!
Luca

**Reply by cescoffier** [reply]
they are hybrid :-D 

The wiring of the component happens at build time, BUT the materialization at runtime.
